drivers/gpu/drm/panel/msm8916-generated/panel-samsung-hx8389c.c

```diff
@@ -0,0 +1,388 @@
+	int ret;
+
+	if (!ctx->prepared)
+		return 0;
+
+	ret = hx8389c_off(ctx);
+	if (ret < 0)
+		dev_err(dev, "Failed to un-initialize panel: %d\n", ret);
+
+	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+	regulator_bulk_disable(ARRAY_SIZE(ctx->supplies), ctx->supplies);
+
+	ctx->prepared = false;
+	return 0;
+}
+
+static const struct drm_display_mode hx8389c_mode = {
+	.clock = (540 + 10 + 14 + 10) * (960 + 9 + 2 + 5) * 60 / 1000,
+	.hdisplay = 540,
+	.hsync_start = 540 + 10,
+	.hsync_end = 540 + 10 + 14,
+	.htotal = 540 + 10 + 14 + 10,
+	.vdisplay = 960,
+	.vsync_start = 960 + 9,
+	.vsync_end = 960 + 9 + 2,
+	.vtotal = 960 + 9 + 2 + 5,
+	.width_mm = 62,
+	.height_mm = 110,
+};
+
+static int hx8389c_get_modes(struct drm_panel *panel,
+			     struct drm_connector *connector)
+{
+	struct drm_display_mode *mode;
+
+	mode = drm_mode_duplicate(connector->dev, &hx8389c_mode);
+	if (!mode)
+		return -ENOMEM;
+
+	drm_mode_set_name(mode);
+
+	mode->type = DRM_MODE_TYPE_DRIVER | DRM_MODE_TYPE_PREFERRED;
+	connector->display_info.width_mm = mode->width_mm;
+	connector->display_info.height_mm = mode->height_mm;
+	drm_mode_probed_add(connector, mode);
+
+	return 1;
+}
+
+static const struct drm_panel_funcs hx8389c_panel_funcs = {
+	.prepare = hx8389c_prepare,
+	.unprepare = hx8389c_unprepare,
+	.get_modes = hx8389c_get_modes,
+};
+
+static int hx8389c_bl_update_status(struct backlight_device *bl)
+{
+	struct mipi_dsi_device *dsi = bl_get_data(bl);
+	u16 brightness = backlight_get_brightness(bl);
+	int ret;
+
+	dsi->mode_flags &= ~MIPI_DSI_MODE_LPM;
+
+	ret = mipi_dsi_dcs_set_display_brightness(dsi, brightness);
+	if (ret < 0)
+		return ret;
+
+	dsi->mode_flags |= MIPI_DSI_MODE_LPM;
+
+	return 0;
+}
+
+// TODO: Check if /sys/class/backlight/.../actual_brightness actually returns
+// correct values. If not, remove this function.
+static int hx8389c_bl_get_brightness(struct backlight_device *bl)
+{
+	struct mipi_dsi_device *dsi = bl_get_data(bl);
+	u16 brightness;
+	int ret;
+
+	dsi->mode_flags &= ~MIPI_DSI_MODE_LPM;
+
+	ret = mipi_dsi_dcs_get_display_brightness(dsi, &brightness);
+	if (ret < 0)
+		return ret;
+
+	dsi->mode_flags |= MIPI_DSI_MODE_LPM;
+
+	return brightness & 0xff;
+}
+
+static const struct backlight_ops hx8389c_bl_ops = {
+	.update_status = hx8389c_bl_update_status,
+	.get_brightness = hx8389c_bl_get_brightness,
+};
+
+static struct backlight_device *
+hx8389c_create_backlight(struct mipi_dsi_device *dsi)
+{
+	struct device *dev = &dsi->dev;
+	const struct backlight_properties props = {
+		.type = BACKLIGHT_RAW,
+		.brightness = 255,
+		.max_brightness = 255,
+	};
+
+	return devm_backlight_device_register(dev, dev_name(dev), dev, dsi,
+					      &hx8389c_bl_ops, &props);
+}
+
+static int hx8389c_probe(struct mipi_dsi_device *dsi)
+{
+	struct device *dev = &dsi->dev;
+	struct hx8389c *ctx;
+	int ret;
+
+	ctx = devm_kzalloc(dev, sizeof(*ctx), GFP_KERNEL);
+	if (!ctx)
+		return -ENOMEM;
+
+	ctx->supplies[0].supply = "vreg";
```

Comment:
```suggestion
	ctx->supplies[0].supply = "vddio";
```

Todo: Update l-p-d script